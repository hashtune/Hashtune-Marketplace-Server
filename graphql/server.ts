import { ApolloServer } from 'apollo-server-express';
// import EthSigUtil from 'eth-sig-util';
import express from 'express';
import { createServer } from 'http';
import passport from 'passport';
import { createContext } from './context';
import { schema } from './schema';
var cors = require('cors');

const cookieParser = require('cookie-parser');

const { PORT = 5000 } = process.env;
const session = require('express-session');

const app = express();

// TODO move to separate file ./auth/passport.ts
// TODO use this passport middleware in mutations
app.use(session({ secret: 'secret', resave: true, saveUninitialized: true }));
app.use(cookieParser());
app.use(passport.initialize());
app.use(passport.session());
app.use(
  cors({
    origin: '*',
  })
);

app.use(express.json());
passport.serializeUser((user, done) => {
  done(null, user);
});
passport.deserializeUser((user: any, done) => {
  done(null, user);
});

const secret = process.env.SERVER_SECRET ?? '';

const JwtStrategy = require('passport-jwt').Strategy;
passport.use(
  new JwtStrategy(
    {
      jwtFromRequest: (req: any) => req.cookies['jwt'] ?? null,
      secretOrKey: secret,
    },
    async function (jwt_payload, done) {
      console.log({ jwt_payload });
      console.log('WEE');
      if (jwt_payload && jwt_payload.user) {
        return done(null, jwt_payload.user);
      } else {
        // user is false if they do not have a valid jwt
        return done(null, false);
      }
    }
  )
);

export const jwtRequired = passport.authenticate('jwt', { session: false });

app.post('/login', (req: any, res: any, next: any) => {
  passport.authenticate('jwt', async (err, user: any) => {
    console.log(req.query);
    const signedMessage = req.query.message;
    const publicKey = req.query.publicKey;

    if (user) {
      // already logged in then redirect to home
      //return next(user);
    }
    return res.json('signature and public key did not match');
  })(req, res, next);
});

app.get('/protected', jwtRequired, (req, res, next) => {
  console.log('approvied');
});

const server = createServer(app);

export const apollo = new ApolloServer({
  schema,
  introspection: true,
  apollo: {},
  context: createContext,
});

export async function main() {
  await apollo.start();
  apollo.applyMiddleware({ app });
  server.listen({ port: PORT }, () => {
    process.stdout.write(
      ` ðŸš€  Server ready at http://localhost:${PORT}${apollo.graphqlPath}`
    );
  });
}
process.env.STAGE != 'test' && main();
