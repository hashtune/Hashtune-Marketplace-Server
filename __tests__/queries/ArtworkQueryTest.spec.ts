import server from '../server';

describe('Test artwork queries', () => {
  const ARTWORK_QUERY = `
    query Query($listArtworksAuction: Boolean) {
        listArtworks(auction: $listArtworksAuction) {
            saleType
        }
    }
`;

  it('should query only auctions ', async () => {
    const res = await server.executeOperation({
      query: ARTWORK_QUERY,
      variables: { listArtworksAuction: true },
    });
    expect(res).toMatchSnapshot();
  });

  it('should query both auctions and fixed', async () => {
    const res = await server.executeOperation({
      query: ARTWORK_QUERY,
      variables: { listArtworksAuction: false },
    });
    expect(res).toMatchSnapshot();
  });
});
