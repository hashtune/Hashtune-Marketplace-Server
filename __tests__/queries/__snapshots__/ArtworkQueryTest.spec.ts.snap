// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test artwork queries should find an artwork by id 1`] = `
Object {
  "data": Object {
    "findArtwork": Object {
      "Artworks": Array [
        Object {
          "title": "amazingsongT1itle",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork queries should never return a pending artwork in a findUnique or findMany query 1`] = `
Object {
  "data": Object {
    "listArtworks": Object {
      "Artworks": Array [
        Object {
          "handle": "4",
          "saleType": "fixed",
        },
        Object {
          "handle": "5",
          "saleType": "fixed",
        },
        Object {
          "handle": "art 1",
          "saleType": "auction",
        },
        Object {
          "handle": "1",
          "saleType": "auction",
        },
        Object {
          "handle": "2",
          "saleType": "auction",
        },
        Object {
          "handle": "3",
          "saleType": "auction",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork queries should not find an artwork by id and throw an error 1`] = `
Object {
  "data": Object {
    "findArtwork": Object {
      "Artworks": null,
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": Object {
        "message": "Couldn't find artwork",
      },
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork queries should query both auctions and fixed 1`] = `
Object {
  "data": Object {
    "listArtworks": Object {
      "Artworks": Array [
        Object {
          "handle": "4",
          "saleType": "fixed",
        },
        Object {
          "handle": "5",
          "saleType": "fixed",
        },
        Object {
          "handle": "art 1",
          "saleType": "auction",
        },
        Object {
          "handle": "1",
          "saleType": "auction",
        },
        Object {
          "handle": "2",
          "saleType": "auction",
        },
        Object {
          "handle": "3",
          "saleType": "auction",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork queries should query only auctions  1`] = `
Object {
  "data": Object {
    "listArtworks": Object {
      "Artworks": Array [
        Object {
          "handle": "3",
          "saleType": "auction",
        },
        Object {
          "handle": "2",
          "saleType": "auction",
        },
        Object {
          "handle": "1",
          "saleType": "auction",
        },
        Object {
          "handle": "art 1",
          "saleType": "auction",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork queries should query only fixed sales  1`] = `
Object {
  "data": Object {
    "listArtworks": Object {
      "Artworks": Array [
        Object {
          "handle": "5",
          "saleType": "fixed",
        },
        Object {
          "handle": "4",
          "saleType": "fixed",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork queries should query only fixed sales that are listed  1`] = `
Object {
  "data": Object {
    "listArtworks": Object {
      "Artworks": Array [
        Object {
          "handle": "5",
          "saleType": "fixed",
        },
        Object {
          "handle": "4",
          "saleType": "fixed",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork queries should return false if a handle is not available 1`] = `
Object {
  "data": Object {
    "handle": false,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork queries should return true if a handle is available 1`] = `
Object {
  "data": Object {
    "handle": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
